AmplifyTests/CategoryTests/Auth/AuthCategoryConfigurationTests.swift

```diff
@@ -0,0 +1,367 @@
+                }
+            }
+        }
+        try Amplify.add(plugin: plugin)
+
+        let config = AuthCategoryConfiguration(
+            plugins: ["MockAuthCategoryPlugin": true]
+        )
+
+        let amplifyConfig = AmplifyConfiguration(auth: config)
+
+        try Amplify.configure(amplifyConfig)
+        try Amplify.Auth.getPlugin(for: "MockAuthCategoryPlugin").configure(using: true)
+        waitForExpectations(timeout: 1.0)
+    }
+
+    /// Test if unconfigured plugin throws an error
+    ///
+    /// - Given: An unconfigured Amplify framework with Auth plugin added
+    /// - When:
+    ///    - Invoke an API in auth
+    /// - Then:
+    ///    - I should get an exception
+    ///
+    func testPreconditionFailureInvokingBeforeConfig() throws {
+        let plugin = MockAuthCategoryPlugin()
+        try Amplify.add(plugin: plugin)
+
+        // Remember, this test must be invoked with a category that doesn't include an Amplify-supplied default plugin
+        let exception: BadInstructionException? = catchBadInstruction {
+            _ = Amplify.Auth.changePassword(currentPassword: "current",
+                                            newPassword: "new",
+                                            listener: nil)
+        }
+        XCTAssertNotNil(exception)
+    }
+
+    // MARK: - Test internal config behavior guarantees
+
+    /// Test if configuring twice throws an exception
+    ///
+    /// - Given: Amplify with Auth plugin configured
+    /// - When:
+    ///    - I try to configure Auth again
+    /// - Then:
+    ///    - Should throw an exception
+    ///
+    func testThrowsConfiguringTwice() throws {
+        let plugin = MockAuthCategoryPlugin()
+        try Amplify.add(plugin: plugin)
+        let config = AuthCategoryConfiguration(
+            plugins: ["MockAuthCategoryPlugin": true]
+        )
+
+        try Amplify.Auth.configure(using: config)
+        XCTAssertThrowsError(try Amplify.Auth.configure(using: config),
+                             "configure() an already configured plugin should throw") { error in
+                                guard case ConfigurationError.amplifyAlreadyConfigured = error else {
+                                    XCTFail("Expected ConfigurationError.amplifyAlreadyConfigured")
+                                    return
+                                }
+        }
+    }
+
+    /// Test if configuring after reset works
+    ///
+    /// - Given: Amplify framework with Auth configured
+    /// - When:
+    ///    - I reset Amplify and then configure again
+    /// - Then:
+    ///    - Should not throw any error
+    ///
+    func testCanConfigureAfterReset() throws {
+        let plugin = MockAuthCategoryPlugin()
+        try Amplify.add(plugin: plugin)
+        let config = AuthCategoryConfiguration(
+            plugins: ["MockAuthCategoryPlugin": true]
+        )
+
+        try Amplify.Auth.configure(using: config)
```

Comment:
1.  I guess technically this could be:
```
XCTAssertNoThrow(try Amplify.Auth.configure...)
```
2.  Also, is there some way to check configure finishes to completion other than just checking for it not throwing?

Reply:
Added for both points.